### ISSUE TITLE
Dynamic imports via ImportManager cripples Python tooling

### ISSUE THREAD

**Issue by kureta** [Abjad/abjad#883]
First of all, thanks for this amazing piece of software.

I am using Pycharm for Python development and it has amazing features that help me write faster and better code. It has auto-complete, it can display method signatures, and it can show documentation right in my code with the press of a key. These features can lower the barrier to entry to coding in Python considerably. Since Abjad is written for composers, if people were able to benefit from these features, or similar features in vim, Emacs, or other editors, many more people would use it in. But since most of the classes are dynamically imported via the ImportManager, these tools do not work. I have to go back and forth between the documentation and Abjad's source code and my code.

So, would you consider abandoning the ImportManager class? I can help with necessary modifications.

Thanks again.

**Comment by josiah-wolf-oberholtzer** [comment]
I understand the point. We've discussed this internally before, but the Abjad developers don't use the kind of tooling you describe and removing the dynamic imports would hamper our work developing the system. So I'm +/- 0 on the issue.

That said, we're preparing for Abjad 3.0, with a number of massive changes. Removing the dynamic imports could be fitted in there, but that work will have to wait for our refactoring to wrap up.

**Comment by trevorbaca** [comment]
Thanks for opening this thread, and for the thoughtful comments. I'm actually sympathetic to this argument for removing the ImportManager. The ImportManager (and its predecessors) have been in the system for almost a decade, and I'd always suspected that at some point the balance of utility would tip from the benefits the ImportManager provided (automatic public import-into-package, regardless of changing package contents) to disadvantages like neutering Pycharm. Hearing that @josiah-wolf-oberholtzer is +/- zero on the topic reinforces that this thinking is probably right.

I suppose the TL;DR here is that ImportManager engaged in magic (of a type that would better be removed), which is great: nice to finally have a motivation to make the change.

It won't take me much effort to make the necessary changes:

- [ ] Rewrite all Abjad `__init__.py` files by hand
- [ ] Remove `ImportManager`

But before I do, it would help if both of you guys (@kureta and @josiah-wolf-oberholtzer) could validate my approach: I'll rewrite initializers that look like ...
```
from abjad.tools import systemtools


systemtools.ImportManager.import_structured_package(
    __path__[0],
    globals(),
    )
```
... to look like ...
```
from .AbjadConfiguration import AbjadConfiguration                              
from .BenchmarkScoreMaker import BenchmarkScoreMaker                            
from .Configuration import Configuration                                        
from .FilesystemState import FilesystemState                                    
from .ForbidUpdate import ForbidUpdate                                          
from .FormatSpecification import FormatSpecification                            
from .IOManager import IOManager                                                
from .ImportManager import ImportManager                                        
from .IndicatorWrapper import IndicatorWrapper                                  
from .LilyPondFormatBundle import LilyPondFormatBundle                          
from .LilyPondFormatManager import LilyPondFormatManager                        
from .NullContextManager import NullContextManager                              
from .ProgressIndicator import ProgressIndicator                                
from .RedirectedStreams import RedirectedStreams                                
from .Signature import Signature                                                
from .SlotContributions import SlotContributions                                
from .StorageFormatAgent import StorageFormatAgent                              
from .StorageFormatSpecification import StorageFormatSpecification              
from .TemporaryDirectory import TemporaryDirectory                              
from .TemporaryDirectoryChange import TemporaryDirectoryChange                  
from .TestManager import TestManager                                            
from .TestCase import TestCase                                                  
from .Timer import Timer                                                        
from .UpdateManager import UpdateManager                                        
from .WellformednessManager import WellformednessManager                        
```
... instead.

(The last example is the current version of `systemtools/__init__.py`; a small handful of `__init__.py` files throughout the system already bypass `ImportManager` in favor of explicit by-hand management of package contents.)

Approach sound ok?

**Comment by kureta** [comment]
Sounds OK to me but I guess @josiah-wolf-oberholtzer is in the process of refactoring the code, so we'll have to wait for him. I made the necessary changes [here](https://github.com/kureta/abjad/tree/static_imports). I can send a pull request if it is OK.

**Comment by josiah-wolf-oberholtzer** [comment]
@kureta I've been waiting on @trevorbaca to merge in the bulk of his recent work before making any changes myself, most of which involve removing code. Once Trevor finishes that up (maybe even done now?) this issue can move forward.

**Comment by trevorbaca** [comment]
Awesome, guys: all my stuff's merged in!

**Comment by kureta** [comment]
I was about to open another issue but I guess this is the right place. Why do we add stuff to Python's builtins? Is that really necessary? That seems like obfuscation/magic too. Even modules as large as numpy are able to do everything they do without adding anything to __builtin__. I can also get rid of that if you guys agree.

**Comment by josiah-wolf-oberholtzer** [comment]
@kureta Re: builtins, it seemed like a convenience in the past (and did make the code read cleanly in its way!) but we've been moving towards namespacing everything with `abjad.` for a while now.

I'm +100 on not putting anything into Python's builtins module. That also goes for the handful of custom exceptions found in Abjad.

Let's split the work up though:
- remove custom exceptions from builtins (see [here](https://github.com/Abjad/abjad/blob/76b538a0267d095c6510093466d31a70164d913c/abjad/__init__.py#L97-L105))
- remove custom constants from builtins (see [here](https://github.com/Abjad/abjad/blob/76b538a0267d095c6510093466d31a70164d913c/abjad/tools/datastructuretools/__init__.py))
- removing dynamic imports

**Comment by kureta** [comment]
I have sent a pull request for static imports. #892 

Could not remove ImportManager class as it is still used for adding custom exceptions into builtins. I'll move on to remove exceptions and constants from builtins, then delete ImportManager.py

**Comment by kureta** [comment]
@josiah-wolf-oberholtzer Why are constants defined twice? In datastructuretools/__init__.py and in abjad/__init__.py

**Comment by trevorbaca** [comment]
No good reason for constants to be defined twice; in fact that situation led to at least one really difficult bug I removed earlier this fall.

**Comment by kureta** [comment]
I have moved all constants into a file named constants within datastructuretools. Also there are tests related to abjad-ide at Travis CI, so it is impossible to pass tests after removing ImportManager without also updating abjad-ide.

One last thing: there are many `import abjad` statements throughout the code. All of that code is import from within abjad, and they also import abjad. This does not feel good. I can also fix those if it's OK for you guys.

One really last thing: I guess I messed up my pull request. Everything is on the same branch, therefore within a single pull request. Sorry for that.

**Comment by trevorbaca** [comment]
The so-called 'late imports' (of `import abjad`) used to even messier than they are now: all such late imports used to be of tools packages within abjad: `from abjad import datastructuretools`, `from abjad import indicatortools`, etc.

Python guidelines frown on late imports, and they've always been something of a thorn in our side. Honestly the reason that they're in the codebase is that we haven't ever figured out a way to remove them. Do you think you see a way that they could actually be gotten rid of? I'd be quite amazed (and happy!) if it were possible to do! (Just beware that there's probably a 90% chance that the problem is much more difficult than it appears.)

I'll remove ImportManager from the IDE; gimme a few minutes ...

**Comment by trevorbaca** [comment]
OK @kureta, ImportManager is removed from the IDE. abjad-ide should be good to go if retest against Travis-CI.

**Comment by kureta** [comment]
I guess this issue can be closed after review. Related pull request passes the tests now.
